🌐 [i18n-KO] Translated `perf_infer_gpu_one.md` to Korean (#24978)
huggingface/transformers: docs/source/ko/perf_infer_gpu_one.md

<!--Copyright 2022 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->

# 단일 GPU에서 효율적인 추론 [[efficient-inference-on-a-single-gpu]]

이 가이드 외에도, [단일 GPU에서의 훈련 가이드](perf_train_gpu_one)와 [CPU에서의 추론 가이드](perf_infer_cpu)에서도 관련 정보를 찾을 수 있습니다.

> ## Better Transformer: PyTorch 네이티브 Transformer 고속 경로 [[better-transformer-pytorchnative-transformer-fastpath]]

Author: ```suggestion
## Better Transformer: PyTorch 네이티브 Transformer 패스트패스 [[better-transformer-pytorchnative-transformer-fastpath]]
```

https://pytorch.org/blog/accelerated-pytorch-2/
https://www.youtube.com/watch?v=CVyggj3Nv-g&t=123

fastpath 관련하여 여러 참고자료들을 살펴보니 Transformer의 attention과 같이 Better Transformer에서 활용되는 일종의 아키텍처인 것으로 짐작됩니다. 어텐션과의 일관성을 위해 패스트패스로 수정하고자 합니다!

> PyTorch 네이티브 [`nn.MultiHeadAttention`](https://pytorch.org/blog/a-better-transformer-for-fast-transformer-encoder-inference/) 어텐션 고속 경로인 BetterTransformer는 [🤗 Optimum 라이브러리](https://huggingface.co/docs/optimum/bettertransformer/overview)의 통합을 통해 Transformers와 함께 사용할 수 있습니다.

Author: ```suggestion
PyTorch 네이티브 [`nn.MultiHeadAttention`](https://pytorch.org/blog/a-better-transformer-for-fast-transformer-encoder-inference/) 어텐션 패스트패스인 BetterTransformer는 [🤗 Optimum 라이브러리](https://huggingface.co/docs/optimum/bettertransformer/overview)의 통합을 통해 Transformers와 함께 사용할 수 있습니다.
```

https://pytorch.org/blog/accelerated-pytorch-2/
https://www.youtube.com/watch?v=CVyggj3Nv-g&t=123

fastpath 관련하여 여러 참고자료들을 살펴보니 Transformer의 attention과 같이 Better Transformer에서 활용되는 일종의 아키텍처인 것으로 짐작됩니다. 어텐션과의 일관성을 위해 패스트패스로 수정하고자 합니다!

> PyTorch의 어텐션 고속 경로는 커널 퓨전과 [중첩된 텐서](https://pytorch.org/docs/stable/nested.html)의 사용을 통해 추론 속도를 높일 수 있습니다. 자세한 벤치마크는 [이 블로그 글](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2)에서 확인할 수 있습니다.

Author: ```suggestion
PyTorch의 어텐션 패스트패스는 커널 퓨전과 [중첩된 텐서](https://pytorch.org/docs/stable/nested.html)의 사용을 통해 추론 속도를 높일 수 있습니다. 자세한 벤치마크는 [이 블로그 글](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2)에서 확인할 수 있습니다.
```

https://pytorch.org/blog/accelerated-pytorch-2/
https://www.youtube.com/watch?v=CVyggj3Nv-g&t=123

fastpath 관련하여 여러 참고자료들을 살펴보니 Transformer의 attention과 같이 Better Transformer에서 활용되는 일종의 아키텍처인 것으로 짐작됩니다. 어텐션과의 일관성을 위해 패스트패스로 수정하고자 합니다!

[`optimum`](https://github.com/huggingface/optimum) 패키지를 설치한 후에는 추론 중 Better Transformer를 사용할 수 있도록 [`~PreTrainedModel.to_bettertransformer`]를 호출하여 관련 내부 모듈을 대체합니다:

```python
model = model.to_bettertransformer()
```

[`~PreTrainedModel.reverse_bettertransformer`] 메소드는 정규화된 transformers 모델링을 사용하기 위해 모델을 저장하기 전 원래의 모델링으로 돌아갈 수 있도록 해줍니다:

```python
model = model.reverse_bettertransformer()
model.save_pretrained("saved_model")
```

> PyTorch 2.0부터는 어텐션 고속 경로가 인코더와 디코더 모두에서 지원됩니다. 지원되는 아키텍처 목록은 [여기](https://huggingface.co/docs/optimum/bettertransformer/overview#supported-models)에서 확인할 수 있습니다.

Author: ```suggestion
PyTorch 2.0부터는 어텐션 패스트패스가 인코더와 디코더 모두에서 지원됩니다. 지원되는 아키텍처 목록은 [여기](https://huggingface.co/docs/optimum/bettertransformer/overview#supported-models)에서 확인할 수 있습니다.
```

https://pytorch.org/blog/accelerated-pytorch-2/
https://www.youtube.com/watch?v=CVyggj3Nv-g&t=123

fastpath 관련하여 여러 참고자료들을 살펴보니 Transformer의 attention과 같이 Better Transformer에서 활용되는 일종의 아키텍처인 것으로 짐작됩니다. 어텐션과의 일관성을 위해 패스트패스로 수정하고자 합니다!

## FP4 혼합 정밀도 추론을 위한 `bitsandbytes` 통합 [[bitsandbytes-integration-for-fp4-mixedprecision-inference]]

> `bitsandbytes`를 설치하고 GPU에서 쉬운 모델 압축의 이점을 얻을 수 있습니다. FP4 양자화를 사용하면 원래의 전체 정밀도 버전과 비교하여 모델 크기를 최대 8배 줄일 수 있습니다. 아래에서 시작하는 방법을 확인하세요.

Author: ```suggestion
`bitsandbytes`를 설치하고 GPU에서 쉬운 모델 압축의 이점을 얻을 수 있습니다. FP4 양자화를 사용하면 원래의 전체 정밀도 버전과 비교하여 모델 크기를 최대 8배 줄일 수 있습니다. 아래에서 방법을 확인하세요.
```

보다 자연스럽게 읽히도록 수정하겠습니다.

<Tip>

이 기능은 다중 GPU 설정에서도 사용할 수 있습니다.

</Tip>

> ### 요구 사항 [[requirements]]

Author: ```suggestion
### 요구 사항 [[requirements-for-fp4-mixedprecision-inference]]
```

올려주신 문서 참고하여 anchor link 수정했습니다! 감사합니다~

```py
model_name = "bigscience/bloom-2b5"
model_4bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_4bit=True)
```
하지만 `accelerate`를 사용하여 각 GPU에 할당할 GPU RAM을 제어할 수 있습니다. 다음과 같이 `max_memory` 인자를 사용하세요:

```py
max_memory_mapping = {0: "600MB", 1: "1GB"}
model_name = "bigscience/bloom-3b"
model_4bit = AutoModelForCausalLM.from_pretrained(
    model_name, device_map="auto", load_in_4bit=True, max_memory=max_memory_mapping
)
```
이 예에서는 첫 번째 GPU가 600MB의 메모리를 사용하고 두 번째 GPU가 1GB를 사용합니다.

### 고급 사용법 [[advanced-usage]]

이 방법의 더 고급 사용법에 대해서는 [양자화](main_classes/quantization) 문서 페이지를 참조하세요.

## Int8 혼합 정밀도 행렬 분해를 위한 `bitsandbytes` 통합 [[bitsandbytes-integration-for-int8-mixedprecision-matrix-decomposition]]

<Tip>

이 기능은 다중 GPU 설정에서도 사용할 수 있습니다.

</Tip>

[`LLM.int8() : 8-bit Matrix Multiplication for Transformers at Scale`](https://arxiv.org/abs/2208.07339) 논문에서 우리는 몇 줄의 코드로 Hub의 모든 모델에 대한 Hugging Face 통합을 지원합니다.
이 방법은 `float16` 및 `bfloat16` 가중치에 대해 `nn.Linear` 크기를 2배로 줄이고, `float32` 가중치에 대해 4배로 줄입니다. 이는 절반 정밀도에서 이상치를 처리함으로써 품질에 거의 영향을 미치지 않습니다.

![HFxbitsandbytes.png](https://s3.amazonaws.com/moonup/production/uploads/1659861207959-62441d1d9fdefb55a0b7d12c.png)

> Int8 혼합 정밀도 행렬 분해는 행렬 곱셈을 두 개의 스트림으로 분리합니다: (1) fp16로 곱해지는 체계적인 특이값 이상치 스트림 행렬(0.01%) 및 (2) int8 행렬 곱셈의 일반적인 스트림(99.9%). 이 방법을 사용하면 매우 큰 모델에 대해 예측적인 저하 없이 int8 추론이 가능합니다.

Author: ```suggestion
Int8 혼합 정밀도 행렬 분해는 행렬 곱셈을 두 개의 스트림으로 분리합니다: (1) fp16로 곱해지는 체계적인 특이값 이상치 스트림 행렬(0.01%) 및 (2) int8 행렬 곱셈의 일반적인 스트림(99.9%). 이 방법을 사용하면 매우 큰 모델에 대해 예측 저하 없이 int8 추론이 가능합니다.
```

자연스럽게 읽히도록 수정하겠습니다.

이 방법에 대한 자세한 내용은 [논문](https://arxiv.org/abs/2208.07339)이나 [통합에 관한 블로그 글](https://huggingface.co/blog/hf-bitsandbytes-integration)에서 확인할 수 있습니다.

> ![MixedInt8.gif](https://s3.amazonaws.com/moonup/production/uploads/1660567469965-62441d1d9fdefb55a0b7d12c.gif)

Reviewer: ```suggestion
![MixedInt8.gif](https://cdn-uploads.huggingface.co/production/uploads/1660567469965-62441d1d9fdefb55a0b7d12c.gif)
```
same here

커널은 GPU 전용으로 컴파일되어 있기 때문에 혼합 8비트 모델을 실행하려면 GPU가 필요합니다. 이 기능을 사용하기 전에 모델의 1/4(또는 모델 가중치가 절반 정밀도인 경우 절반)을 저장할 충분한 GPU 메모리가 있는지 확인하세요.
이 모듈을 사용하는 데 도움이 되는 몇 가지 참고 사항이 아래에 나와 있습니다. 또는 [Google colab](#colab-demos)에서 데모를 따라할 수도 있습니다.

> ### 요구 사항 [[requirements]]

Author: ```suggestion
### 요구 사항 [[requirements-for-int8-mixedprecision-matrix-decomposition]]]]
```

51 line과 함께 anchor link 오류를 해결하기 위한 수정입니다!

> ### 요구 사항 [[requirements]]

Author: ```suggestion
### 요구 사항 [[requirements-for-int8-mixedprecision-matrix-decomposition]]
```

- `bitsandbytes<0.37.0`을 사용하는 경우, 8비트 텐서 코어(Turing, Ampere 또는 이후 아키텍처 - 예: T4, RTX20s RTX30s, A40-A100)를 지원하는 NVIDIA GPU에서 실행하는지 확인하세요. `bitsandbytes>=0.37.0`을 사용하는 경우, 모든 GPU가 지원됩니다.
- 올바른 버전의 `bitsandbytes`를 다음 명령으로 설치하세요:
`pip install bitsandbytes>=0.31.5`
- `accelerate`를 설치하세요
`pip install accelerate>=0.12.0`

### 혼합 Int8 모델 실행 - 단일 GPU 설정 [[running-mixedint8-models-single-gpu-setup]]

> 필요한 라이브러리를 설치한 후 혼합 8비트 모델을 로드하는 방법은 다음과 같습니다:

Reviewer 2: ```suggestion
필요한 라이브러리를 설치한 후 혼합 8비트 모델을 가져오는 방법은 다음과 같습니다:
```
glossary에 기반하여 수정해 보았습니다.

Author: 감사합니다!

* 입력을 모델과 동일한 GPU에 배치하는 것이 좋습니다.

다음은 간단한 예입니다:

```py
from transformers import AutoModelForCausalLM, AutoTokenizer

model_name = "bigscience/bloom-2b5"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model_8bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_8bit=True)

prompt = "Hello, my llama is cute"
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
generated_ids = model.generate(**inputs)
outputs = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
```

### 혼합 Int8 모델 실행 - 다중 GPU 설정 [[running-mixedint8-models-multi-gpu-setup]]

다중 GPU에서 혼합 8비트 모델을 로드하는 방법은 단일 GPU 설정과 동일합니다(동일한 명령어 사용):
```py
model_name = "bigscience/bloom-2b5"
model_8bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_8bit=True)
```
하지만 `accelerate`를 사용하여 각 GPU에 할당할 GPU RAM을 제어할 수 있습니다. 다음과 같이 `max_memory` 인자를 사용하세요:

```py
max_memory_mapping = {0: "1GB", 1: "2GB"}
model_name = "bigscience/bloom-3b"
model_8bit = AutoModelForCausalLM.from_pretrained(
    model_name, device_map="auto", load_in_8bit=True, max_memory=max_memory_mapping
)
```

> 이 예에서는 첫 번째 GPU가 1GB의 메모리를 사용하고 두 번째 GPU가 2GB를 사용합니다.

Reviewer 2: ```suggestion
이 예시에서는 첫 번째 GPU가 1GB의 메모리를 사용하고 두 번째 GPU가 2GB를 사용합니다.
```

조금 더 자연스러운 느낌으로 수정해 보았습니다.

PyTorch 네이티브 [`nn.MultiHeadAttention`](https://pytorch.org/blog/a-better-transformer-for-fast-transformer-encoder-inference/) 어텐션 고속 경로인 BetterTransformer는 [🤗 Optimum 라이브러리](https://huggingface.co/docs/optimum/bettertransformer/overview)의 통합을 통해 Transformers와 함께 사용할 수 있습니다.

PyTorch의 어텐션 고속 경로는 커널 퓨전과 [중첩된 텐서](https://pytorch.org/docs/stable/nested.html)의 사용을 통해 추론 속도를 높일 수 있습니다. 자세한 벤치마크는 [이 블로그 글](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2)에서 확인할 수 있습니다.

[`optimum`](https://github.com/huggingface/optimum) 패키지를 설치한 후에는 추론 중 Better Transformer를 사용하기 위해 관련 내부 모듈을 호출하여 대체합니다.:

```python
model = model.to_bettertransformer()
```

[`~PreTrainedModel.reverse_bettertransformer`] 메소드는 정규화된 transformers 모델링을 사용하기 위해 모델을 저장하기 전에 원래의 모델링으로 돌아갈 수 있도록 해줍니다.:

```python
model = model.reverse_bettertransformer()
model.save_pretrained("saved_model")
```

PyTorch 2.0부터는 어텐션 고속 경로가 인코더와 디코더 모두에서 지원됩니다. 지원되는 아키텍처 목록은 [여기](https://huggingface.co/docs/optimum/bettertransformer/overview#supported-models)에서 확인할 수 있습니다.

## FP4 혼합 정밀도 추론을 위한 `bitsandbytes` 통합 [[bitsandbytes-integration-for-fp4-mixedprecision-inference]]

> `bitsandbytes`를 설치하고 GPU에서 쉬운 모델 압축의 이점을 얻을 수 있습니다. FP4 양자화를 사용하면 원래의 전체 정밀도 버전과 비교하여 모델 크기를 최대 8배 줄일 수 있습니다. 아래에서 시작하는 방법을 확인하세요.

Author: 훨씬 자연스러운 흐름이 된 것 같습니다! 감사합니다!

<Tip>

이 기능은 다중 GPU 설정에서도 사용할 수 있습니다.

</Tip>

> ### 요구 사항 [[requirements]]

Author: 영문 원본 문서 수정을 위해 몇가지 질문이 있어 코멘트를 남깁니다!

1. 현재 제 깃허브에 올라온 main과 허깅페이스의 main 간 commit 차이가 있는데 이를 PR로 당겨온 뒤 새로운 브랜치를 판 뒤 수정하는 것이 현재 브랜치에서 진행하는 것보다 더 나은 방법일지 궁금합니다!

2. 영문 원본 문서 수정에 있어서도 리뷰어로 저희 팀원들을 멘션해도 괜찮을까요?

> ### 요구 사항 [[requirements]]

Reviewer 3: 1. 
> 현재 제 깃허브에 올라온 main과 허깅페이스의 main 간 commit 차이가 있는데 이를 PR로 당겨온 뒤 새로운 브랜치를 판 뒤 수정하는 것이 현재 브랜치에서 진행하는 것보다 더 나은 방법일지 궁금합니다!
- 네, PR별로 브랜치를 따로 사용하시는 것이 좋습니다. 새 PR을 열어 영어 원본을 수정하신다면 말씀하신 방법으로 새 브랜치를 만들어 진행해주세요!
2. 
> 영문 원본 문서 수정에 있어서도 리뷰어로 저희 팀원들을 멘션해도 괜찮을까요?
- 번역 이슈가 아니라서 팀원들을 멘션하지 않으셔도 괜찮습니다. 물론 수정한 부분에 대해 의견을 구하고 싶다면 멘션을 해주셔도 문제 없습니다!
 
문의 주신 내용과 관련하여 참고하실 수 있는 이슈가 하나 더 있더라구요. 원본 문서 수정 건만 따로 올린 PR이라 이슈/PR 본문을 쓰시거나 브랜치 이름을 정하실 때 보다 참고가 되실 것 같아 추가로 공유 드립니다:
- 이슈: https://github.com/huggingface/transformers/issues/23729

- 최신 `bitsandbytes` 라이브러리
`pip install bitsandbytes>=0.39.0`

- 최신 `accelerate`를 소스에서 설치
`pip install git+https://github.com/huggingface/accelerate.git`

- 최신 `transformers`를 소스에서 설치
`pip install git+https://github.com/huggingface/transformers.git`

### FP4 모델 실행 - 단일 GPU 설정 - 빠른 시작 [[running-fp4-models-single-gpu-setup-quickstart]]

다음 코드를 실행하여 단일 GPU에서 빠르게 FP4 모델을 실행할 수 있습니다.

```py
from transformers import AutoModelForCausalLM

model_name = "bigscience/bloom-2b5"
model_4bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_4bit=True)
```
`device_map`은 선택 사항입니다. 그러나 `device_map = 'auto'`로 설정하는 것이 사용 가능한 리소스를 효율적으로 디스패치하기 때문에 추론에 있어 권장됩니다.

### FP4 모델 실행 - 다중 GPU 설정 [[running-fp4-models-multi-gpu-setup]]

> 다중 GPU에서 혼합 4비트 모델을 로드하는 방법은 단일 GPU 설정과 동일합니다(동일한 명령어 사용):

Reviewer 3: ```suggestion
다중 GPU에서 혼합 4비트 모델을 가져오는 방법은 단일 GPU 설정과 동일합니다(동일한 명령어 사용):
```
glossary에 기반하여 수정 제안을 드립니다!

```py
model_name = "bigscience/bloom-2b5"
model_4bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_4bit=True)
```

> 하지만 `accelerate`를 사용하여 각 GPU에 할당할 GPU RAM을 제어할 수 있습니다. 다음과 같이 `max_memory` 인자를 사용하세요:

Reviewer 3: ```suggestion
하지만 `accelerate`를 사용하여 각 GPU에 할당할 GPU RAM을 제어할 수 있습니다. 다음과 같이 `max_memory` 인수를 사용하세요:
```
glossary에 기반한 수정 제안입니다.

Author: 놓친 부분이 있었네요 감사합니다!

model_8bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_8bit=True)
```

텍스트 생성의 경우:

> * `pipeline()` 함수 대신 모델의 `generate()` 메소드를 사용하는 것을 권장합니다. `pipeline()` 함수로는 추론이 가능하지만, 혼합 8비트 모델에 최적화되지 않았기 때문에 `generate()` 메소드를 사용하는 것보다 느릴 수 있습니다. 또한, nucleus 샘플링과 같은 일부 샘플링 전략은 혼합 8비트 모델에 대해 `pipeline()` 함수에서 지원되지 않습니다.

Reviewer 3: 원문에 `nucleaus`라고 오타가 있어서 한글 문서에서는 정정해주셨군요! 👍

```

### 혼합 Int8 모델 실행 - 다중 GPU 설정 [[running-mixedint8-models-multi-gpu-setup]]

다중 GPU에서 혼합 8비트 모델을 로드하는 방법은 단일 GPU 설정과 동일합니다(동일한 명령어 사용):
```py
model_name = "bigscience/bloom-2b5"
model_8bit = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", load_in_8bit=True)
```

> 하지만 `accelerate`를 사용하여 각 GPU에 할당할 GPU RAM을 제어할 수 있습니다. 다음과 같이 `max_memory` 인자를 사용하세요:

Reviewer 3: ```suggestion
하지만 `accelerate`를 사용하여 각 GPU에 할당할 GPU RAM을 제어할 수 있습니다. 다음과 같이 `max_memory` 인수를 사용하세요:
```
glossary에 따른 수정 제안입니다.

    model_name, device_map="auto", load_in_8bit=True, max_memory=max_memory_mapping
)
```
이 예에서는 첫 번째 GPU가 1GB의 메모리를 사용하고 두 번째 GPU가 2GB를 사용합니다.

### Colab 데모 [[colab-demos]]

이 방법을 사용하면 이전에 Google Colab에서 추론할 수 없었던 모델에 대해 추론할 수 있습니다.

> Google Colab에서 T5-11b(42GB in fp32)를 실행하는 데모를 확인하세요. Google Colab에서 8비트 양자화를 사용:

Reviewer 3: ```suggestion
Google Colab에서 8비트 양자화를 사용하여 T5-11b(42GB in fp32)를 실행하는 데모를 확인하세요:
```
끝나는 문장을 자연스럽게 해보고자 두 문장을 이어보았는데, 어떠실까요?

[![Open In Colab: T5-11b demo](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1YORPWx4okIHXnjW7MSAidXN29mPVNT7F?usp=sharing)

> 또는 BLOOM-3B에 대한 데모:

Reviewer 3: ```suggestion
또는 BLOOM-3B에 대한 데모를 확인하세요:
```
이 줄도 전체적인 문장 톤에 맞추어 보았습니다.
